### PR TITLE
fix: handle timestamp[ns] cast failure in data preview

### LINKED ISSUE #19: Data preview crashes on tables with nanosecond-precision timestamps
## Description

The **Data** tab preview fails when viewing Delta tables whose parquet files contain `timestamp[ns]` columns. PyArrow refuses to cast `timestamp[ns]` to `timestamp[us, tz=UTC]` when the nanosecond values cannot be losslessly represented in microseconds.

## Error

```
❌ Data preview failed: Casting from timestamp[ns] to timestamp[us, tz=UTC] would lose data: -48521116231933723624
```

## Affected code paths

| Path | Method | Impact |
|------|--------|--------|
| `onelake_client/tables/delta.py` | `DeltaTableReader.read_sample()` | Delta Data tab preview |
| `onelake_tui/detail.py` | `_read_parquet_fallback()` | Fallback parquet preview |
| `onelake_tui/detail.py` | `_preview_parquet()` | Direct parquet file preview |

## Root cause

`DeltaTableReader.read_sample()` calls `dt.to_pyarrow_dataset().head(limit)`, which internally attempts to cast `timestamp[ns]` columns to `timestamp[us, tz=UTC]`. PyArrow's default `safe=True` mode rejects this when precision would be lost — particularly for corrupt or sentinel values like `-48521116231933723624`.

## Expected behavior

Data preview should render the table, with out-of-range timestamp values displayed as `null` rather than crashing the entire preview.

## Steps to reproduce

1. Open OneLake TUI and navigate to a workspace with a lakehouse containing a Delta table with `timestamp[ns]` columns (e.g. `dbo/dimension_city`)
2. Select the table in the tree
3. Click **Load Data Preview** in the Data tab
4. Observe the error

## Environment

- OneLake TUI v0.3.0
- deltalake 1.5.0
- pyarrow 23.0.1

### PR DESCRIPTION
**Pull request by lmoloney**
## Summary

Fixes #19 — Data preview crashes on Delta tables whose parquet files contain `timestamp[ns]` columns.

## Problem

`DeltaTableReader.read_sample()` calls `dt.to_pyarrow_dataset().head(limit)`, which internally casts `timestamp[ns]` → `timestamp[us, tz=UTC]` with `safe=True`. PyArrow rejects this for values that cannot be losslessly represented in microseconds (e.g. values like `-4852116231933723624` whose nanosecond component is not divisible by 1000), producing:

```
❌ Data preview failed: Casting from timestamp[ns] to timestamp[us, tz=UTC] would lose data
```

### 1. ArrowInvalid fallback in `read_sample()`
When `ds.head()` raises `ArrowInvalid` for the timestamp cast failure, falls back to reading parquet fragments using their physical schema (which preserves `timestamp[ns]`). Limited to 10 fragments for performance safety.

### 2. `coerce_timestamps()` public helper
Scans a pyarrow Table for `timestamp[ns]` columns and casts them to `timestamp[us]` with `safe=False` (truncates sub-microsecond precision). Applied to all three read paths:
- `DeltaTableReader.read_sample()` — Delta Data tab preview
- `_read_parquet_fallback()` — fallback parquet preview
- `_preview_parquet()` — direct parquet file preview

### 3. `_nullify_out_of_range()` defensive guard
After casting, timestamps outside year 0001–9999 are replaced with null. For the ns→us path this is defensive (all int64 ns values map to years 1677–2262), but provides a safety net for future use.

## Test coverage

341 tests pass. New tests include:

**Unit tests (TestCoerceTimestamps — 9 tests):** ns→us cast, timezone preservation, MagicMock passthrough, pre-existing nulls, in-range survival

**Unit tests (TestNullifyOutOfRange — 7 tests):** boundary values (year 0001, 9999), extreme int64 values, mixed valid/invalid, timezone handling

**Integration tests (TestReadSample — 7 tests):** happy path, limit, empty, coercion, ArrowInvalid fallback with real fragment mocks, multi-fragment limit, non-timestamp error propagation

**End-to-end tests (TestTimestampCoercionE2E — 4 tests):** Real parquet files written with timestamp[ns], read through timestamp[us] dataset schema, the exact production value `-4852116231933723624`, mixed rows

## Also included

- CI job enforcing `CHANGELOG.md` updates on user-facing PRs (skip with `chore` or `documentation` label)
- Code-review instructions (`.github/instructions/code-review.instructions.md`)
- Copilot instructions updated with timestamp coercion convention

## Checklist

- [x] All 341 tests pass
- [x] Lint + format clean (`ruff check` + `ruff format`)
- [x] CHANGELOG updated
- [x] Verified against real Fabric data — preview loads correctly

## Summary by CodeRabbit

* **New Features**
  * Added a public timestamp-coercion helper to make data previews resilient to ns→us timestamp mismatches.

* **Bug Fixes**
  * Fixed crashes when previewing tables with nanosecond timestamps by downcasting and nullifying out-of-range values; added a safe fallback read path.

* **Chores**
  * Added CI enforcement requiring changelog entries for user-facing PRs.

* **Documentation**
  * Updated contributor instructions to require changelog and docs updates for user-facing changes.

* **Tests**
  * Added comprehensive tests covering coercion and preview fallback behavior.

### CODE REVIEW

File: TUI/tests/test_delta_reader.py [hunk 1]
```diff
@@ -533,6 +534,69 @@ def test_error_result_raises(self):
...
+
+        # Value that fits in int64 but overflows when cast ns→us with safe=True
+        corrupt_val = -(2**62)
+        arr = pa.array([corrupt_val], type=pa.timestamp("ns"))
+        table = pa.table({"ts": arr})
+        # safe=False means this won't raise
+        result = _coerce_timestamps(table)
+        assert result.num_rows == 1
+        assert result.schema.field("ts").type == pa.timestamp("us")
+
```
[the remark this answers is not in the document]
Comment (lmoloney) [reply TUI/tests/test_delta_reader.py]: @copilot please review these findings

File: TUI/src/onelake_tui/detail.py [hunk 2]
```diff
@@ -29,6 +29,7 @@
 
 from onelake_client import OneLakeClient
 from onelake_client.exceptions import FileTooLargeError
+from onelake_client.tables.delta import _coerce_timestamps
```
[the remark this answers is not in the document]
Comment (lmoloney) [reply TUI/src/onelake_tui/detail.py]: @copilot apply changes based on [this feedback](https://github.com/lmoloney/OneLakeTools/pull/20#discussion_r3124403858)

File: TUI/src/onelake_client/tables/delta.py [hunk 3]
```diff
@@ -283,7 +322,10 @@ async def read_sample(
 
         def _head():
             ds = dt.to_pyarrow_dataset()
-            return ds.head(limit)
+            table = ds.head(limit)
+            if not hasattr(table, "num_columns") or not hasattr(table, "schema"):
+                return table
+            return coerce_timestamps(table)
```
[the remark this answers is not in the document]
Comment (lmoloney) [reply TUI/src/onelake_client/tables/delta.py]: @copilot please fix
Verify each finding against the current code and only fix it if needed.

In `@TUI/src/onelake_client/tables/delta.py` around lines 325 - 328, Add a
regression test in the TestReadSample suite that constructs a real pyarrow.Table
with one or more timestamp('ns') columns and actual values, then calls the same
code path exercised by read_sample (so it reaches the branch that executes
ds.head(...) and returns coerce_timestamps(table)); assert the returned table's
schema uses timestamp('us') for those columns and that the data values were
properly converted (e.g., timestamps rounded/truncated to microsecond
precision). Use coerce_timestamps and read_sample identifiers to locate the path
under test and ensure you don't mock ds.head so the real coercion logic runs
end-to-end.

**Comment by lmoloney** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/lmoloney/OneLakeTools/pull/20#pullrequestreview-4161248198)

**Comment by lmoloney** [comment]
@copilot 

Verify each finding against the current code and only fix it if needed.

Nitpick comments:
In `@TUI/tests/test_delta_reader.py`:
- Around line 576-577: Rename the test function
test_corrupt_value_does_not_crash to reflect that it validates precision-loss
truncation for in-range nanosecond values (e.g.,
test_in_range_ns_precision_truncation_or_safe_truncation_does_not_crash) and
update its docstring to describe that in-range ns values that lose precision
with safe=True are truncated safely rather than being nullified; ensure all
references to the old name (if any) are updated accordingly.

**Comment by lmoloney** [comment]
All 4 open review threads addressed in d38d62c:

**Thread 7** (hasattr guard bypassed by MagicMock): Replaced with `isinstance(table, pa.Table)` check inside `coerce_timestamps()`. Non-pyarrow inputs pass through unchanged. Removed the `hasattr` guard from `read_sample()`.

**Thread 8** (CI changelog 2-dot diff): Changed to triple-dot (`$BASE...$HEAD`) for merge-base-aware comparison.

**Thread 13 & 14** (detail.py coerce_timestamps with MagicMock): The `isinstance` guard inside `coerce_timestamps()` handles these — MagicMock inputs return unchanged without needing guards at call sites.

**Additional changes:**
- Added `_nullify_out_of_range()` helper: after `safe=False` cast, timestamps outside year 0001–9999 are replaced with null
- 10 new tests covering nullification boundaries, extreme values, isinstance passthrough, and mixed valid/invalid data
- Fixed CHANGELOG: corrected the null description, added entries for `coerce_timestamps` public API, CI enforcement, and code-review instructions

All 333 tests pass, lint clean.

**Comment by lmoloney** [comment]
@coderabbitai review

**Comment by lmoloney** [comment]
@coderabbitai full review

**Comment by lmoloney** [comment]
**Re: nitpick — `pa.ArrowInvalid` vs `pa.lib.ArrowInvalid`**

Checked: `pa.ArrowInvalid is pa.lib.ArrowInvalid` → they are the same object. Both work identically.

Keeping `pa.lib.ArrowInvalid` for now because:
- It's what pyarrow itself raises (the exception class lives in `pyarrow.lib`)
- It's consistent with the production catch in `delta.py:366`
- Both forms are stable — pyarrow re-exports lib exceptions to the top-level namespace

If pyarrow ever removes the `lib` re-export (unlikely — it's the canonical module), a single find-replace fixes it. Not worth churn on this PR.
